XenAdmin/Core/VMGroup.cs

```diff
@@ -158,7 +158,13 @@ internal static string UpsellLearnMoreUrl
             get { return typeof(T) == typeof(VMSS) ? InvisibleMessages.UPSELL_LEARNMOREURL_VMSS : InvisibleMessages.UPSELL_LEARNMOREURL_VM_APPLIANCES; }
         }
 
-        internal static AsyncAction AssignVMsToGroupAction(T group, List<XenRef<VM>> vms, bool suppressHistory)
+        internal static AsyncAction AssignVMsToGroupAction(T group, List<XenRef<VM>> vmRefs, bool suppressHistory)

```

Comment:
You're sometimes passing XenRef<VM>s and sometimes VMs, and we keep parallel functions/constructors several levels deep. It would be worth seeing whether you can resolve the references earlier and always just pass VMs.